### PR TITLE
Issue 225

### PR DESCRIPTION
**Pull request by vqd8a**
Address #225

### CODE REVIEW

**Comment by vqd8a** [comment]
Added arith traits overload for vector complex types in [KokkosBatched_Vector.hpp](https://github.com/kokkos/kokkos-kernels/blob/issue-225/src/batched/KokkosBatched_Vector.hpp) 
Modified [Test_Batched_VectorArithmatic.hpp](https://github.com/kokkos/kokkos-kernels/blob/issue-225/unit_test/batched/Test_Batched_VectorArithmatic.hpp) respectively.
Still leave RealPart() in [KokkosBatched_Vector.hpp](https://github.com/kokkos/kokkos-kernels/blob/issue-225/src/batched/KokkosBatched_Vector.hpp) since I am not sure where it is called.

**Comment by vqd8a** [comment]
Thanks, @kyungjoo-kim.
Couldn't these be taken care by imag() in ArithTraits? Why did we need ImagPart for these cases?
```C++
template<typename T> KOKKOS_FORCEINLINE_FUNCTION typename Kokkos::Details::ArithTraits<T>::mag_type ImagPart(const T &val) { return 0; }
template<typename T> KOKKOS_FORCEINLINE_FUNCTION typename Kokkos::Details::ArithTraits<Kokkos::complex<T> >::mag_type ImagPart(const Kokkos::complex<T> &val) { return val.imag(); }
```
I will change RealPart and ImagPart used in other routines too.

**Comment by kyungjoo-kim** [comment]
if you added traits, it is not necessary. they are just quick solution when I wrote first time.

**Comment by vqd8a** [comment]
Changed RealPart and ImagPart used in other routines to real() and imag().

**Comment by vqd8a** [comment]
Thanks @mhoemmen . The code is fixed.

**Comment by kyungjoo-kim** [comment]
@vqd8a Yhis needs to wait until #232 is merged first.

**Comment by vqd8a** [comment]
Thanks, @kyungjoo-kim . Could you please help answer @mhoemmen 's questions about MagnitudeScalarType and ValueScalarType?

**Comment by kyungjoo-kim** [comment]
@vqd8a They are probably a bi-product when I develop the code. It is a quick solution without thinking too much ;). I don't think that I use it in the code but we can change the code as @mhoemmen suggests.

**Comment by mhoemmen** [comment]
@kyungjoo-kim PURGE THEM :D

**Comment by vqd8a** [comment]
Since ValueScalarType is not used elsewhere, I just removed it.
As @mhoemmen suggested, I also changed MagnitudeScalarType to ArithTraits<T>::mag_type. Other routines that used MagnitudeScalarType were also changed accordingly.

File: src/batched/KokkosBatched_LU_Decl.hpp [hunk 1]
```diff
@@ -16,7 +16,7 @@ namespace KokkosBatched {
       KOKKOS_INLINE_FUNCTION
       static int
       invoke(const AViewType &A,
-             const typename MagnitudeScalarType<typename AViewType::non_const_value_type>::type tiny = 0);
+             const typename Kokkos::Details::ArithTraits<typename AViewType::non_const_value_type>::mag_type tiny = 0);
     };
```
Comment (kyungjoo-kim) [inline src/batched/KokkosBatched_LU_Decl.hpp]: This is wrong. In the batched routines, input value_type can be two -  double or SIMD<double> type. When SIMD<double> type is used, your modification returns SIMD<mag_type<double> > but the previous routine MagnitudeScalarType returns double type.

Comment (vqd8a) [reply src/batched/KokkosBatched_LU_Decl.hpp]: Should we keep ```MagnitudeScalarType```?

Comment (kyungjoo-kim) [reply src/batched/KokkosBatched_LU_Decl.hpp]: You have two options: keep MagnitudeScalarType or using the ValueScalarType
```
typename Kokkos::Details::ArithTraits<typename ValueScalarType<typename AViewType::non_const_value_type>::type>::mag_type
```

Comment (vqd8a) [reply src/batched/KokkosBatched_LU_Decl.hpp]: I think we should go back to MagnitudeScalarType

**Comment by vqd8a** [comment]
@kyungjoo-kim Please review these changes.

**Comment by kyungjoo-kim** [comment]
BTW, did you check the code compiles and pass all the unit tests ?

**Comment by vqd8a** [comment]
thanks @kyungjoo-kim . Yes, it passed all the unit tests.

**Comment by kyungjoo-kim** [comment]
thank you for fixing the code.

**Comment by crtrott** [comment]
Passed tests other than the clang 6 error in SPGEMM_KK_MEMORY:
```
#######################################################
PASSED TESTS
#######################################################
clang-3.9.0-Pthread_Serial-release build_time=225 run_time=1702
cuda-9.1-Cuda_OpenMP-release build_time=619 run_time=1091
gcc-4.8.4-OpenMP-release build_time=138 run_time=317
gcc-4.8.4-Pthread-release build_time=125 run_time=879
gcc-5.3.0-Serial-release build_time=153 run_time=763
intel-16.0.1-OpenMP-release build_time=403 run_time=374
#######################################################
FAILED TESTS
#######################################################
clang-6.0-Cuda_Pthread-release (test failed)
```

File: src/batched/KokkosBatched_Vector.hpp [hunk 2]
```diff
@@ -138,13 +138,13 @@ namespace Kokkos {
 	
 	
     template<typename T, int l>
-    class ArithTraits<Vector<SIMD<Kokkos::complex<T>>,l> > { 
+    class ArithTraits<KokkosBatched::Experimental::Vector<KokkosBatched::Experimental::SIMD<Kokkos::complex<T>>,l> > {
```
Comment (mhoemmen) [inline src/batched/KokkosBatched_Vector.hpp]: No love for my template template, alas...

Comment (crtrott) [reply src/batched/KokkosBatched_Vector.hpp]: Uhm that didn't compile without the change ...

Comment (vqd8a) [reply src/batched/KokkosBatched_Vector.hpp]: when I looked at the old code RealPart and ImagPart, I saw they were only for SIMD. That's why I did not try @mhoemmen 's solution template template.
Now, I have just tried it and it works. Should I create a new PR for this?

Comment (mhoemmen) [reply src/batched/KokkosBatched_Vector.hpp]: @vqd8a I don't mean to tell you what to do :-)  If it needs to work for AVX, etc. as well as SIMD, then my approach would give you a generic solution.

Comment (vqd8a) [reply src/batched/KokkosBatched_Vector.hpp]: @crtrott @kyungjoo-kim should we need this generic solution?

Comment (mhoemmen) [reply src/batched/KokkosBatched_Vector.hpp]: (You don't _need_ it, but you do need ArithTraits, etc. to work for `AVX<T>` and any other `SimdImplementation<T>` you want to support, as well as `SIMD<T>`.  However that happens is up to you :-) .)

**Comment by kyungjoo-kim** [comment]
right now I unified avx into simd. but I still open the possibility that someone else wants to try this kind of structure with a different tag that can guide to a different instruction set e.g., ARM. I think that the current solution is fine and let's not worry about things not happend yet.

**Pull request merged**